### PR TITLE
feat: support string delimiter and quote

### LINKED ISSUE #640: Support delimiter with multiple characters when importing/exporting CSV
We may support delimiter with multiple characters for CSV format.

Refer to Spark's docs https://spark.apache.org/docs/latest/sql-data-sources-csv.html .

image

### LINKED ISSUE #661: Support quote in FileOptionsParser options
Mysql and spark can distinguish "" and the separator can be part of the value. We can support this feature for correctly reading some input data such as "aa", "bb".
image

### PR DESCRIPTION
**Pull request by zekai427**
- Resolve：https://github.com/4paradigm/OpenMLDB/issues/661
- Resolve：https://github.com/4paradigm/OpenMLDB/issues/640
- Docs for internal developer：https://wiki.4paradigm.com/pages/viewpage.action?pageId=106578099

### CODE REVIEW

File: src/cmd/split.h [hunk 1]
```diff
@@ -91,69 +92,72 @@ char* strdup_with_new(const char* the_string) {
...
             }
         }
-        const bool need_another_column = (*line == delimiter) && (line == end_of_line - 1);
         *end = '\0';
         cols->push_back(start);
         // If line was something like [paul,] (comma is the last character
         // and is not proceeded by whitespace or quote) then we are about
         // to eliminate the last column (which is empty). This would be
         // incorrect.
+        const bool need_another_column = (line + delimiter_len == end_of_line);
```
Comment (tobegit3hub) [inline src/cmd/split.h:136]: We should check if `*line` is the first char of the delimiter to handle the case of the comment above.

**Review by tobegit3hub — CHANGES_REQUESTED** [review]
Please add some unit tests to cover the new options such as "enclosed" and "string delimiter".

File: src/cmd/split.h [hunk 2]
```diff
@@ -91,69 +92,72 @@ char* strdup_with_new(const char* the_string) {
...
         if (need_another_column) cols->push_back(end);
 
-        assert(*line == '\0' || *line == delimiter);
+        assert(*line == '\0' || *line == delimiter[0]);
```
Comment (tobegit3hub) [inline src/cmd/split.h]: compare with `memcmp`.

**Comment by zekai427** [comment]
> Please add some unit tests to cover the new options such as "enclosed" and "string delimiter".

Done

File: src/cmd/split.h [hunk 3]
```diff
@@ -91,69 +92,76 @@ char* strdup_with_new(const char* the_string) {
...
     char* end;
     char* start;
+    int delimiter_len = strlen(delimiter);
 
-    for (; line < end_of_line; line++) {
-        // Skip leading whitespace, unless said whitespace is the delimiter.
-        while (absl::ascii_isspace(*line) && *line != delimiter) ++line;
+    for (; line < end_of_line; line += delimiter_len) {
+        // Skip leading whitespace, unless said whitespace is the part of delimiter.
+        while (absl::ascii_isspace(*line) && *line != delimiter[0]) ++line;
```
Comment (vagetablechicken) [inline src/cmd/split.h:104]: if delimter is empty, it will coredump here.
prevent it.

Comment (vagetablechicken) [reply src/cmd/split.h:104]: add test about 'space is delimiter'

Comment (zekai427) [reply src/cmd/split.h:104]: Resolved

File: src/cmd/split.h [hunk 4]
```diff
@@ -91,69 +92,76 @@ char* strdup_with_new(const char* the_string) {
 // See //util/csv/parser.h for more complete documentation.
 //
 // ----------------------------------------------------------------------
-void SplitCSVLineWithDelimiter(char* line, char delimiter, std::vector<char*>* cols) {
```
Comment (jingchen2222) [inline src/cmd/split.h:94]: Add unit test for SplitLineWithDelimiter.
- empty line
- delimiter `' '`, `','`, `'---'`, `''`, etc

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Review by vagetablechicken — APPROVED** [review]

File: src/cmd/sql_cmd_test.cc [hunk 5]
```diff
@@ -210,21 +231,123 @@ TEST_F(SqlCmdTest, deploy) {
...
+    std::vector<std::string> cols;
+    openmldb::cmd::SplitLineWithDelimiterForStrings(" --- --- ---", "---", &cols, '\0');
+    ASSERT_EQ(cols.size(), 4);
+    for (auto& line : cols) {
+        ASSERT_EQ(strcmp(line.c_str(), ""), 0);
+    }
+    cols.clear();
+
+    openmldb::cmd::SplitLineWithDelimiterForStrings("- - ---ab---c b---", "---", &cols, '\0');
+    ASSERT_EQ(cols.size(), 4);
```
Comment (dl239) [inline src/cmd/sql_cmd_test.cc]: move these cases to split_test.cc

Comment (zekai427) [reply src/cmd/sql_cmd_test.cc]: Resolved

**Comment by lumianph** [comment]
[bugfix tracking](https://trello.com/c/wEQ9HW0k/25-bugfix-tracking)

**Pull request merged**